apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/TimedBufferingConsumer.cs

```diff
@@ -61,38 +51,58 @@ private TimedBufferingConsumer(IConsumer<T> consumer, TimeSpan waitTime, IClock
         /// <param name="waitTime">The minimum amount of time between automatic flushes.</param>
         /// <param name="clock">A clock for getting the current timestamp.</param>
         public static TimedBufferingConsumer<T> Create(IConsumer<T> consumer, TimeSpan waitTime, IClock clock = null)
-        {
-            return new TimedBufferingConsumer<T>(consumer, waitTime, clock ?? SystemClock.Instance);
-        }
+            => new TimedBufferingConsumer<T>(consumer, waitTime, clock ?? SystemClock.Instance);
 
         /// <inheritdoc />
-        public void Receive(IEnumerable<T> items)
+        public override void Receive(IEnumerable<T> items)
         {
             GaxPreconditions.CheckNotNull(items, nameof(items));
-            lock (_mutex)
+            Semaphore.Wait();
+            try
             {
-                _items.AddRange(items);
-                if (_clock.GetCurrentDateTimeUtc() >= _nextFlush) { 
-                  Flush();
+                Items.AddRange(items);
+                if (_clock.GetCurrentDateTimeUtc() >= _nextFlush)
+                {
+                    FlushWithSemaphoreHeld();
                 }
             }
+            finally
+            {
+                Semaphore.Release();
+            }
         }
 
         /// <inheritdoc />
-        public void Flush()
+        public override async Task ReceiveAsync(IEnumerable<T> items, CancellationToken cancellationToken = default(CancellationToken))
         {
-            IList<T> old;
-            lock (_mutex)
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            await Semaphore.WaitAsync(cancellationToken);
+            try
             {
-                old = _items;
-                _items = new List<T>();
-                _nextFlush = DateTime.UtcNow.Add(_waitTime);
+                Items.AddRange(items);
+                if (_clock.GetCurrentDateTimeUtc() >= _nextFlush)
+                {
+                    await FlushAsyncWithSemaphoreHeld(cancellationToken);
+                }
             }
-
-            if (old.Count != 0)
+            finally
             {
-                _consumer.Receive(old);
+                Semaphore.Release();
             }
         }
+
+        /// <inheritdoc />
+        protected override void FlushWithSemaphoreHeld()
+        {
+            _nextFlush = DateTime.UtcNow.Add(_waitTime);
+            base.FlushWithSemaphoreHeld();
+        }
+
+        /// <inheritdoc />
+        protected override async Task FlushAsyncWithSemaphoreHeld(CancellationToken cancellationToken)
+        {
+            _nextFlush = DateTime.UtcNow.Add(_waitTime);
+            await base.FlushAsyncWithSemaphoreHeld(cancellationToken);
```

Comment:
Remove `async` and `await` and just return here. No sense in making the compiler generate all the async machinery here.